core/planner.go

```diff
@@ -0,0 +1,158 @@
+package core
+
+import (
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/benbjohnson/clock"
+	"github.com/evcc-io/evcc/api"
+	"github.com/evcc-io/evcc/util"
+)
+
+const (
+	chargeEfficiency   = 0.95
+	hysteresisDuration = 5 * time.Minute
+)
+
+// RatesByPrice implements sort.Interface based on price
+type RatesByPrice []api.Rate
+
+func (a RatesByPrice) Len() int {
+	return len(a)
+}
+
+func (a RatesByPrice) Less(i, j int) bool {
+	if a[i].Price == a[j].Price {
+		return a[i].Start.After(a[j].Start)
+	}
+	return a[i].Price < a[j].Price
+}
+
+func (a RatesByPrice) Swap(i, j int) {
+	a[i], a[j] = a[j], a[i]
+}
+
+type Planner struct {
+	log    *util.Logger
+	clock  clock.Clock // mockable time
+	tariff api.Tariff
+	mux    sync.Mutex
+}
+
+func NewPlanner(log *util.Logger, tariff api.Tariff) *Planner {
+	clock := clock.New()
+	return &Planner{
+		log:    log,
+		clock:  clock,
+		tariff: tariff,
+	}
+}
+
+func (t *Planner) isCheapSlotNow(duration time.Duration, end time.Time) (bool, error) {
+	cheapSlotNow := false
+	cntExpectedSlots := 0
+	curSlotNr := 0
+	var pi api.Rate
+	var sum time.Duration
+
+	data, err := t.tariff.Rates()
+	if err != nil {
+		return false, err
+	}
+	last := data[len(data)-1].End
+
+	duration = time.Duration(float64(duration) / chargeEfficiency)
+
+	// Save same duration until next price info update
+	if end.After(last) {
+		duration_old := duration
+		duration = time.Duration(float64(duration) * float64(time.Until(last)) / float64(time.Until(end)))
+		t.log.DEBUG.Printf("reduced duration from %s to %s until got new priceinfo after %s\n", duration_old.Round(time.Minute), duration.Round(time.Minute), last.Round(time.Minute))
+	}
+
+	t.log.DEBUG.Printf("charge duration: %s, end: %v, find best prices:\n", duration.Round(time.Minute), end.Round(time.Second))
+
+	sort.Sort(RatesByPrice(data))
+
+	for i := 0; i < len(data); i++ {
+		pi = data[i]
+
+		if pi.Start.Before(t.clock.Now()) && pi.End.After(t.clock.Now()) { // current slot
+			pi.Start = t.clock.Now()
+		}
+
+		if pi.End.Before(t.clock.Now()) { // old data
+			continue
+		}
+
+		if !(pi.Start.Before(end)) { // charge should ends before
```

Comment:
if pi.Start.After(end) || pi.End.Before(t.clock.Now()) { ...

Reply:
> @schenlap was fehlt denn noch um auf "Ready for review" zu kommen?

Ich denke wir sind schon mitten im Review :-), ich schalte um.

Reply:
> if pi.Start.After(end) 

damit hatte ich Probleme da die Zeit während der Berechnung weiter fortschreitet.
After: ist ein >
!Before ist ein < bzw. >=, also nicht das selbe

Zusammenfassen in ein if geht natürlich